RFC: White Flag (#5)
iotaledger/protocol-rfcs: text/0005-white-flag/0005-white-flag.md

deterministic ordering of the tangle and applying only the first bundle(s) that does not violate the ledger state.

The content of this RFC is based on [Conflict white flag: Mitigate conflict spamming by ignoring conflicts](https://iota.cafe/t/conflict-white-flag-mitigate-conflict-spamming-by-ignoring-conflicts/233).

# Motivation

<!-- TODO -->

The main motivations:

- Defend against censorship attacks - Conflicts will no longer block bundles from being approved.
- Make reattachments unnecessary - As long as the network is not saturated, theoretically all bundles should be
approved. And no bundle will be left behind.
- Increase TPS - Due to easy node tip selection the network throughput should increase.
- Increase CTPS - Due to the above, increase in TPS and no left-behinds, we expect CTPS to increase as well.

# Detailed design

<!-- TODO -->

Definitions:
- confirm
- ignore
- reference / indirectly

Let's define a conflicting bundle as a bundle that leads to a negative balance on an address if applied to the current
ledger state.

In case of conflicting bundles with White Flag, nodes will confirm only one bundle and ignore all the others. For this

> to work, they need to be sure they are all confirming the same bundle; hence, the need for a deterministic ordering

Reviewer: ```suggestion
to work, nodes need to be sure they are all confirming the same bundle; hence, the need for a deterministic ordering
```

of the tangle.

## Deterministic ordering of the tangle

<!-- TODO -->

The deterministic ordering of the past-cone of a milestone is a [topological ordering](https://en.wikipedia.org/wiki/Topological_sorting)
generated by a post-order Depth-First Search by going first through trunks, then branches and finally root.

> The stopping condition of this traversal is reaching bundles that are already confirmed.

Reviewer: but what is the starting point of the DFS?
i guess the new milestone?
also, it's a bit confusing to say "The stopping condition of this traversal is reaching bundles that are already confirmed."
because all the transactions in this cone are confirmed, it's just that we stop if we reach transactions that were already confirmed by previous milestones

Author: Right, already changed that in a future commit, thanks.

Once a bundle is marked as ignored, this is final and can't be changed by a later milestone.

Since the ledger state is maintained from one milestone to another, a bundle conflicting with a bundle already confirmed
by a previous milestone would also obviously be ignored.

## Pseudo-code

The following algorithm describes the process of updating the ledger state which is usually triggered by the arrival of
a new milestone confirming many new bundles.

Even though the tangle is a graph made of transactions, for the sake of this pseudo-code it is considered as a graph of
bundles. For this reason, when the `trunk` and/or the `branch` of a bundle are mentioned, they are actually referring to
the `trunk` and/or `branch` of the last transaction of the bundle.

This algorithm is to be followed with care when implementing a node to avoid differences in the ledger state.

```
update_ledger_state(ledger, milestone) {
    s = new Stack()
    visited = new Set()

    s.push(milestone)

    while (!s.empty()) {
        curr = s.peek()

        // only apply if you reached a leaf or both approvees have been visited or confirmed by another milestone

>         if ((curr.trunk == null || visited.contains(curr.trunk) || curr.trunk.snapshot_index != milestone.index)

Reviewer 2: using `milestone.index` s confusing for me, because I don't understand where it is updated...
I would rather use the pseudo-function `trunk.isMarkedByMilestone(milestone)` or `isConfirmed`...

>         if ((curr.trunk == null || visited.contains(curr.trunk) || curr.trunk.snapshot_index != milestone.index)

Reviewer 2: I just realized that `curr.trunk == null` is a mistake

Because if it is null it may mean that we have a non-solid subtangle...
What the check here should be is if `curr.trunk == genesis`

Theoretically, `curr.trunk` should never be null because we solidify the milestone before applying it...
Maybe we should point this out in the RFC?

Reviewer 2: if it is null then there should be an error I believe, because we have holes in the tangle

Author: Sure, I even wondered if it should be an SEP check instead of a genesis check but let's not over complicate it.

Reviewer 2: well yes actually
SEP is like genesis

          && ((curr.branch == null) || visited.contains(curr.branch) || curr.branch.snapshot_index != milestone.index)) {
            ledger.apply(curr)
            visited.add(curr)
            s.pop()
        }

>         else if (curr.trunk != null && !visited.contains(curr.trunk) && curr.trunk.snapshot_index == milestone.index) {

Reviewer 2: ditto

            s.push(curr.trunk)
        }

>         else if (curr.branch != null && !visited.contains(curr.branch) curr.branch.snapshot_index == milestone.index) {

Reviewer 2: ditto

network before [Coordicide](https://coordicide.iota.org/) is complete.

The feature presented in this RFC, White Flag, allows milestones to confirm conflicting bundles by enforcing
deterministic ordering of the Tangle and applying only the first bundle(s) that does not violate the ledger state.

The content of this RFC is based on [Conflict white flag: Mitigate conflict spamming by ignoring conflicts](https://iota.cafe/t/conflict-white-flag-mitigate-conflict-spamming-by-ignoring-conflicts/233).

# Motivation

- Eliminates the [Conflict spamming attack](https://iota.cafe/t/conflict-spamming-attack/232);
- As conflicts are ignored in the balance computation, they don’t need to be considered during tip selection of the
nodes allowing much easier tip selection algorithms leading to increased TPS;
- By using this approach in combination with an appropriate TSA, during regular use, no honest transaction will ever
require re-attaching leading to increased CTPS;
- Doesn't come with added computation complexity by integrating nicely into already existing algorithms;

# Detailed design

First, let's define what it means for a bundle to be:

> - referenced (indirectly or not) by another bundle: it is in the past cone of this bundle;

Reviewer 3: When talking about relations between two or more bundles I would use labels A, B, ... to increase clarity.

+ Feature name: `white-flag`
+ Start date: 2020-03-06
+ RFC PR: [iotaledger/protocol-rfcs#0005](https://github.com/iotaledger/protocol-rfcs/pull/5)

# Summary

This RFC is part of a set of protocol changes, [Chrysalis](https://roadmap.iota.org/chrysalis), aiming at improving the
network before [Coordicide](https://coordicide.iota.org/) is complete.

The feature presented in this RFC, White Flag, allows milestones to confirm conflicting bundles by enforcing
deterministic ordering of the Tangle and applying only the first bundle(s) that does not violate the ledger state.

The content of this RFC is based on [Conflict white flag: Mitigate conflict spamming by ignoring conflicts](https://iota.cafe/t/conflict-white-flag-mitigate-conflict-spamming-by-ignoring-conflicts/233).

# Motivation

- Eliminates the [Conflict spamming attack](https://iota.cafe/t/conflict-spamming-attack/232);

> - As conflicts are ignored in the balance computation, they don’t need to be considered during tip selection of the

Reviewer 4: ```suggestion
- As conflicts are ignored in the balance computation, they do not need to be considered during tip selection of the
```

nodes allowing much easier tip selection algorithms leading to increased TPS;
- By using this approach in combination with an appropriate TSA, during regular use, no honest transaction will ever
require re-attaching leading to increased CTPS;

> - Doesn't come with added computation complexity by integrating nicely into already existing algorithms;

Reviewer 4: ```suggestion
- Does not come with added computation complexity by integrating nicely into already existing algorithms;
```

# Detailed design

> First, let's define what it means for a bundle to be:

Reviewer 4: ```suggestion
First, let us define what it means for a bundle to be:
```

> First, let's define what it means for a bundle to be:
> 
> - referenced (indirectly or not) by another bundle: it is in the past cone of this bundle;
> - confirmed: it is referenced by a milestone;
> - applied: it is confirmed and applied to the ledger state;
> - ignored: it is confirmed but not applied;
> - conflicting: it would lead to an invalid ledger state if applied;

Reviewer 4: ```suggestion
First, let's define what it means for a bundle A to be:

- referenced (indirectly or directly) by bundle B: A is contained in the past cone of B;
- confirmed: A is referenced by a milestone;
- applied: A is confirmed and applied to the ledger state;
- ignored: A is confirmed but not applied;
- conflicting: A would lead to an invalid ledger state if applied;
```
I agree with @Alex6323, it think the above would be a tiny bit clearer.

> In case of conflicting bundles with White Flag, a node applies only one bundle to the ledger state and ignore all the

Reviewer 5: ```suggestion
In case of conflicting bundles with White Flag, a node applies only one bundle to the ledger state and ignores all the
```

others. For this to work, all the nodes need to be sure they are all applying the same bundle; hence, the need for a
deterministic ordering of the Tangle.

> First, this RFC propose a deterministic ordering of the Tangle, then it explain which bundle is selected in case of

Reviewer 5: ```suggestion
First, this RFC proposes a deterministic ordering of the Tangle, then it explains which bundle is selected in case of
```

> conflict.

Reviewer 5: ```suggestion
conflicts.
```

**Note: this RFC is about ledger computation only. For this reason, it assumes that the past cone of a milestone has
already been confirmed and all the referenced bundles have been validated. However, in the event that an invalid bundle
was encountered in the confirmation traversal, a node's expected behaviour would be to completely stop operations - and

> log the error - as it would mean the coordinator confirmed something it shouldn't have.**

Reviewer 4: ```suggestion
log the error - as it would mean the coordinator confirmed something it should not have.**
```

it respects the dependency of bundles, ensuring that the `trunk` and `branch` of a bundle are applied before it. Since
there are multiple valid topological orders for the same graph and in order to avoid conflicting ledger states it is
required that all nodes apply bundles in the exact same order.

For this reason, this RFC propose an order that has to be rigorously followed by all node implementations. This order is
the topological ordering generated by a post-order [Depth-First Search (DFS)](https://en.wikipedia.org/wiki/Depth-first_search)
starting from a milestone and by going first through `trunk` bundle, then `branch` bundle and finally current bundles.
Since only a subset of bundles is considered, the stopping condition of this DFS is reaching bundles that are already
confirmed by another milestone.

## Applying first bundle(s) that does not violate the ledger state

If a conflict is occurring in the set of bundles confirmed by a milestone, nodes have to apply the first - with regards
to the order previously proposed - of the conflicting bundles to the ledger and ignore all the others.

> Once a bundle is marked as ignored, this is final and can't be changed by a later milestone.

Reviewer 4: ```suggestion
Once a bundle is marked as ignored, this is final and cannot be changed by a later milestone.
```

        if ((solid_entry_points.contains(curr.trunk) || visited.contains(curr.trunk) || curr.trunk.confirmation_index != milestone.index)
          && (solid_entry_points.contains(curr.branch) || visited.contains(curr.branch) || curr.branch.confirmation_index != milestone.index)) {
            ledger.apply(curr)
            visited.add(curr)
            s.pop()
        }
        else if (!solid_entry_points.contains(curr.trunk)
          && !visited.contains(curr.trunk)
          && curr.trunk.confirmation_index == milestone.index) {
            s.push(curr.trunk)
        }
        else if (!solid_entry_points.contains(curr.branch)
          && !visited.contains(curr.branch)
          && curr.branch.confirmation_index == milestone.index) {
            s.push(curr.branch)
        }
    }
}
```

**Notes**:
- Even though the tangle is a graph made of transactions, for the sake of this pseudo-code it is considered as a graph
of bundles. For this reason, when the `trunk` and/or the `branch` of a bundle are mentioned, they are actually referring
to the `trunk` and/or `branch` of the last transaction of the bundle.

> - `solid_entry_points` is a set of hashes that are considered solid even though we don't have them or their past in

Reviewer 4: ```suggestion
- `solid_entry_points` is a set of hashes that are considered solid even though we do not have them or their past in
```

+ Feature name: `white-flag`
+ Start date: 2020-03-06
+ RFC PR: [iotaledger/protocol-rfcs#0005](https://github.com/iotaledger/protocol-rfcs/pull/5)

# Summary

> Allow milestones to confirm conflicting bundles. Mutate the ledger state by deterministically ordering the bundles and

Author: Added, thanks

ignoring conflicts.

# Motivation

The main motivations:

- Defend against censorship attacks - Conflicts will no longer block bundles from being approved.
- Make reattachments unnecessary - As long as the network is not saturated, theoretically all bundles should be
approved. And no bundle will be left behind.
- Increase TPS - Due to easy node tip selection the network throughput should increase.
- Increase CTPS - Due to the above, increase in TPS and no left-behinds, we expect CTPS to increase as well.

# Detailed design

Let's define a conflicting bundle as a bundle that leads to a negative balance on an address if applied to the current
ledger state.

When a milestone is accepted by a node then the following algorithm to mutate the ledger state is performed:

> - Look for milestone whose index is the current latest solid milestone index + 1. Milestones must be applied in order!

Author: I removed this part.

- When the milestone is found start traversing down the tangle in DFS going first to the trunk and only then the branch.
Stop at the transactions that has only approved parents.
- Start applying bundles to the diffmap only as you go up the recursion stack (when you climb up back from the leafs).
- Every time you attempt to apply a bundle to the diffmap:
If it is invalid, mark it as seen and ignored by the milestone.
If it is valid and not conflicting with the current state then apply it to the ledger state. Mark it as seen and

> approved.

Reviewer: approved or confirmed?
or maybe a different word because either one it eventually means that the mentioned bundle spent address (and spent into) are deducted/added in the ledger state
